internals-js/src/operations.ts

```diff
@@ -1629,13 +1641,44 @@ export class FieldSelection extends Freezable<FieldSelection> {
         //   }
         // To do that, we can change that `equals` to `contains`, but then we should also "extract" the remainder
         // of `optimizedSelection` that isn't covered by the fragment, and that is the part slighly more involved.
-        if (optimizedSelection.equals(candidate.selectionSet)) {
+        if (optimizedSelection && optimizedSelection.equals(candidate.selectionSet)) {
           const fragmentSelection = new FragmentSpreadSelection(fieldBaseType, fragments, candidate.name);
           return new FieldSelection(this.field, selectionSetOf(fieldBaseType, fragmentSelection));
+        } else if(options?.autoFragmetize && optimizedSelection && optimizedSelection.contains(candidate.selectionSet)) {
+          // As mentioned in the above comment, we find any fragments that match some of the members of the optimizedSelection
+          // If yes, repace part of the optimizedSelection with the existing fragment
+          const fragmentedSelection = new FragmentSpreadSelection(fieldBaseType, fragments, candidate.name);
+          // Get the remainder unmatched slections
+          const selectionSetNotFragmented = optimizedSelection.minus(candidate.selectionSet);
+          // Create a selection set that includes partly matched fragment and unmatched selections
+          const modifiedSelection = new SelectionSet(fieldBaseType, fragments);
+          modifiedSelection.add(fragmentedSelection);
+          if(!selectionSetNotFragmented.isEmpty()) {
+            for(const selection of selectionSetNotFragmented.selections()) {
+              modifiedSelection.add(selection);
+            }
+          }
+          // reassign optimizedSelection
+          if(modifiedSelection)
+            optimizedSelection = modifiedSelection;
+          // In the above partial selection, if the input query contains fragments inside a fragment for a selectionset
+          // we will not be able to regenerate exactly the same because the Map is not ordered.
+          // But defeinitely helps in many cases.
         }
       }
+      // Create a new named fragment with exact selection set, only if we did not find any matching fragment in the input query.
+      // Incase we see this same pattern in another part of the query,
+      // this will help reduce query expansion when sending requests to subgraphs.
+      // If we don't see the same pattern the final pass of optimizer will re-expand it.
+      const schema = options?.schema;
+      if(options?.autoFragmetize && schema && optimizedSelection.selections().length > 1) {
+        const hash = createHash('sha256').update(optimizedSelection.toString()).digest('hex');
```

Comment:
I believe the only goal of that `hash` is to generate a unique name for the created fragment. If so, I would suggest using a simple uuid here (using the `uuidv1()` method already imported in this file). This is going to be substantially cheaper to generate (`optimizedSelection.toString()` is not exactly cheap) and will lead to overall smaller identifiers (more readable, and I guess makes the overall query smaller).